scripts/smb-enum-services.nse

```diff
@@ -0,0 +1,1144 @@
+-- <elem key="wait_hint">0</elem>
+-- </table>
+-- </table>
+-- <table>
+-- <elem key="displayName">Volume Shadow Copy</elem>
+-- <elem key="serviceName">VSS</elem>
+-- <table key="serviceStatus">
+-- <table key="state">
+-- <elem>SERVICE_STATE_ACTIVE</elem>
+-- <elem>SERVICE_STATE_ALL</elem>
+-- </table>
+-- <elem key="win32_exit_code">0</elem>
+-- <elem key="service_exit_code">0</elem>
+-- <table key="controls_accepted">
+-- </table>
+-- <table key="type">
+-- <elem>SERVICE_TYPE_WIN32_OWN_PROCESS</elem>
+-- <elem>SERVICE_TYPE_WIN32</elem>
+-- </table>
+-- <elem key="check_point">0</elem>
+-- <elem key="wait_hint">0</elem>
+-- </table>
+-- </table>
+-- <table>
+-- <elem key="displayName">WMI Performance Adapter</elem>
+-- <elem key="serviceName">WmiApSrv</elem>
+-- <table key="serviceStatus">
+-- <table key="state">
+-- <elem>SERVICE_STATE_ACTIVE</elem>
+-- <elem>SERVICE_STATE_ALL</elem>
+-- </table>
+-- <elem key="win32_exit_code">1077</elem>
+-- <elem key="service_exit_code">0</elem>
+-- <table key="controls_accepted">
+-- </table>
+-- <table key="type">
+-- <elem>SERVICE_TYPE_WIN32_OWN_PROCESS</elem>
+-- <elem>SERVICE_TYPE_WIN32</elem>
+-- </table>
+-- <elem key="check_point">0</elem>
+-- <elem key="wait_hint">0</elem>
+-- </table>
+-- </table>
+
+author = "Rewanth Cool"
+license = "Same as Nmap--See https://nmap.org/book/man-legal.html"
+categories = {"discovery","intrusive","safe"}
+
+portrule = shortport.port_or_service({445, 139}, "microsoft-ds", "tcp", "open")
+
+action = function(host, port)
+
+  local status, smbstate = msrpc.start_smb(host, msrpc.SVCCTL_PATH)
+  status, bind_result = msrpc.bind(smbstate, msrpc.SVCCTL_UUID, msrpc.SVCCTL_VERSION, nil)
+
+  if(status == false) then
+    smb.stop(smbstate)
+    return false, bind_result
+  end
+
+  -- Open the service manager
+  stdnse.debug2("Opening the remote service manager")
+
+  status, open_result = msrpc.svcctl_openscmanagerw(smbstate, host.ip)
+
+  if(status == false) then
+    smb.stop(smbstate)
+    return false, open_result
+  end
+
+
+  --@param dwservicetype The type of services to be enumerated.
+  --                     Lookup table for dwservicetype is as follows:
+  --                       SERVICE_DRIVER - 0x0000000B
+  --                       SERVICE_FILE_SYSTEM_DRIVER - 0x00000002
+  --                       SERVICE_KERNEL_DRIVER - 0x00000001
+  --                       SERVICE_WIN32 - 0x00000030
+  --                       SERVICE_WIN32_OWN_PROCESS - 0x00000010 (default)
+  --                       SERVICE_WIN32_SHARE_PROCESS - 0x00000020
+  local dwservicetype = 0x00000010
+
+  --@param dwservicestate The state of the services to be enumerated.
+  --                      Lookup table for dwservicetype is as follows:
+  --                      SERVICE_ACTIVE - 0x00000001
+  --                      SERVICE_INACTIVE - 0x00000002
+  --                      SERVICE_STATE_ALL - 0x00000003 (default)
+  local dwservicestate = 0x00000003
```

Comment:
Let's make the default SERVICE_ACTIVE. This will reduce output somewhat and also makes it more interesting to look at, since the set of running services can vary more than the set of available services depending on system configuration. Later, we can consider making this a script-arg.